go/tools/macrobench/macrobench.go

```diff
@@ -0,0 +1,138 @@
+/*
+ *
+ * Copyright 2021 The Vitess Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * /
+ */
+
+package macrobench
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/vitessio/arewefastyet/go/mysql"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+const (
+	ErrorNoSysBenchResult          = "no sysbench results were found"
+
+	prefixMacroBenchSysbenchConfig = "macrobench_"
+)
+
+func buildSysbenchArgString(m map[string]string, step string) []string {
+	output := map[string]string{}
+	for k, v := range m {
+		idx := strings.Index(k, "_")
+		if idx < 0 {
+			continue
+		}
+		head := k[:idx]
+		tail := k[idx+1:]
+		if head == step || head == "all" {
+			if _, exists := output[tail]; exists {
+				if head == "all" {
+					continue
+				}
+			}
+			output[tail] = v
+		}
+	}
+
+	var results []string
+	for k, v := range output {
+		results = append(results, fmt.Sprintf("--%s=%s", k, v))
+	}
+	return results
+}
+
+// MacroBench executes a macro benchmark by using sysbench.
+// Based on the given MacroBenchConfig, the function will
+// parse the configuration to send down to sysbench (size of tables
+// duration of benchmark, mysql targets, etc...).
+// After the execution, the output of the last step (stepRun) is
+// converted to a slice of MacroBenchmarkResult, which is then
+// uploaded to MySQL using the mysql.ConfigDB in MacroBenchConfig.
+//
+// We use two forks of sysbench, one for oltp workloads
+// and the other for tpcc workload. We use these forks because
+// they implement a custom method to print results in JSON.
+//
+// Regular Sysbench: https://github.com/planetscale/sysbench
+// Sysbench-TPCC: https://github.com/planetscale/sysbench-tpcc
+func MacroBench(mabcfg MacroBenchConfig) error {
+	var err error
+	var sqlClient *mysql.Client
+
+	if mabcfg.DatabaseConfig != nil && mabcfg.DatabaseConfig.IsValid() {
+		sqlClient, err = mysql.New(*mabcfg.DatabaseConfig)
+		if err != nil {
+			return err
+		}
+		defer sqlClient.Close()
+	}
```

Comment:
I think instead of this. If the condition fails then you should throw an error and then you would not be needed to check later for `sqlclient != nil`

Reply:
I do not want to restrict the use of SQL. The user must be able to run a macro benchmark without necessarily saving its results to a database. If `mabcfg.DatabaseConfig == nil` then `sqlclient` stays `nil` and no error is thrown.